Screenshotter fails after a JavaScript exception

### What?

The screenshotter fails when called in a `catch` block (after an exception).

```javascript
export async function someTest() {
  ...
  try {
    ...
  } catch (error) {
    await page.screenshot(...);
  } finally {
    await page.close();
  }
```

Exception:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x15e1f48]

goroutine 11479469 [running]:
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End.deferwrap1()
go.opentelemetry.io/otel/sdk@v1.29.0/trace/span.go:398 +0x25
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0xc00ff77d40, {0x0, 0x0, 0xc0615c1a40?})
go.opentelemetry.io/otel/sdk@v1.29.0/trace/span.go:436 +0xa62
panic({0x197cc60?, 0x2f63d50?})
runtime/panic.go:785 +0x132
github.com/grafana/xk6-browser/common.(*screenshotter).screenshot(0xc017fa5ee0, {0x1f9b8c0, 0xc01df2c780}, 0x0, 0xc017fa5cd8, {0x1c15435, 0x3}, 0x0, 0xc04cc80d80?, {0xc09641fa10, ...})
github.com/grafana/xk6-browser@v1.9.1/common/screenshotter.go:185 +0x328
github.com/grafana/xk6-browser/common.(*screenshotter).screenshotPage(0xc017fa5ee0, 0xc03747c680, 0xc04cc80ac0)
github.com/grafana/xk6-browser@v1.9.1/common/screenshotter.go:398 +0x334
github.com/grafana/xk6-browser/common.(*Page).Screenshot(0xc03747c680, 0xc04cc80ac0, {0x7f691c1821d0, 0xc000451d10})
github.com/grafana/xk6-browser@v1.9.1/common/page.go:1356 +0x2c6
github.com/grafana/xk6-browser/browser.mapPage.func33.1()
github.com/grafana/xk6-browser@v1.9.1/browser/page_mapping.go:241 +0x4e
github.com/grafana/xk6-browser/k6ext.promise.func1()
github.com/grafana/xk6-browser@v1.9.1/k6ext/promise.go:24 +0x2c
created by github.com/grafana/xk6-browser/k6ext.promise in goroutine 730
github.com/grafana/xk6-browser@v1.9.1/k6ext/promise.go:23 +0x9a
```

Test Run ID: 3439289

### Why?

Users should take screenshots even after an exception to see the issues in their web pages or scripts.

### How?

Looking at the code:

* `screenshotPage` [calls](https://github.com/grafana/xk6-browser/blob/56449e93e2ab48bb2916e3c0f882d33718d3d975/common/screenshotter.go#L398) `screenshotter.screenshot` with a non-nil viewport `*Rect`.
* However, the code panics [here](https://github.com/grafana/xk6-browser/blob/56449e93e2ab48bb2916e3c0f882d33718d3d975/common/screenshotter.go#L185) while calculating `Width:  viewport.Width / visualViewport.Scale`.

Since `viewport` should not be nil, `visualViewport` might be.

### Tasks

```[tasklist]
### Tasks
- [ ] Subissue
- [ ] PR
- [ ] Docs PR
- [ ] TypeScript Definitions PR
- [ ] Update the k6 release notes and release tasks
```


### Related PR(s)/Issue(s)

_No response_

--- a/common/element_handle.go
+++ b/common/element_handle.go
@@ -1317,7 +1317,7 @@ func (h *ElementHandle) Screenshot(
 
 	span.SetAttributes(attribute.String("screenshot.path", opts.Path))
 
-	s := newScreenshotter(spanCtx, sp)
+	s := newScreenshotter(spanCtx, sp, h.logger)
 	buf, err := s.screenshotElement(h, opts)
 	if err != nil {
 		err := fmt.Errorf("taking screenshot of elementHandle: %w", err)

--- a/common/page.go
+++ b/common/page.go
@@ -1327,7 +1327,7 @@ func (p *Page) Screenshot(opts *PageScreenshotOptions, sp ScreenshotPersister) (
 
 	span.SetAttributes(attribute.String("screenshot.path", opts.Path))
 
-	s := newScreenshotter(spanCtx, sp)
+	s := newScreenshotter(spanCtx, sp, p.logger)
 	buf, err := s.screenshotPage(p, opts)
 	if err != nil {
 		err := fmt.Errorf("taking screenshot of page: %w", err)

--- a/common/screenshotter.go
+++ b/common/screenshotter.go
@@ -150,7 +150,6 @@ func (s *screenshotter) restoreViewport(p *Page, originalViewport *Size) error {
 	return p.resetViewport()
 }
 
-//nolint:funlen
 func (s *screenshotter) screenshot(
 	sess session, doc, viewport *Rect, format ImageFormat, omitBackground bool, quality int64, path string,
 ) ([]byte, error) {
@@ -178,30 +177,9 @@ func (s *screenshotter) screenshot(
 		capture.WithFormat(cdppage.CaptureScreenshotFormatPng)
 	}
 
-	// Add clip region
-	//nolint:dogsled
-	_, _, _, _, visualViewport, _, err := cdppage.GetLayoutMetrics().Do(cdp.WithExecutor(s.ctx, sess))
+	visualViewportScale, visualViewportPageX, visualViewportPageY, err := getViewPortDimensions(s.ctx, sess, s.logger)
 	if err != nil {
-		return nil, fmt.Errorf("getting layout metrics for screenshot: %w", err)
-	}
-
-	visualViewportScale := 1.0
-	visualViewportPageX, visualViewportPageY := 0.0, 0.0
-	// we had a null pointer panic cases, when visualViewport is nil
-	// instead of the erroring out, we fallback to defaults and still try to do a screenshot
-	if visualViewport != nil {
-		visualViewportScale = visualViewport.Scale
-		visualViewportPageX = visualViewport.PageX
-		visualViewportPageY = visualViewport.PageY
-	} else {
-		s.logger.Warnf(
-			"Screenshotter::screenshot",
-			"chrome browser returned nil on page.getLayoutMetrics, falling back to defaults for visualViewport "+
-				"(scale: %v, pageX: %v, pageY: %v)."+
-				"This is non-standard behavior, if possible please report this issue (with reproducible script) "+
-				"to the https://github.com/grafana/xk6-browser/issues/1502.",
-			visualViewportScale, visualViewportPageX, visualViewportPageY,
-		)
+		return nil, err
 	}
 
 	if doc == nil {
@@ -253,6 +231,42 @@ func (s *screenshotter) screenshot(
 	}
 
 	return buf, nil
+}
+
+func getViewPortDimensions(ctx context.Context, sess session, logger *log.Logger) (float64, float64, float64, error) {
+	visualViewportScale := 1.0
+	visualViewportPageX, visualViewportPageY := 0.0, 0.0
+
+	// Add clip region
+	//nolint:dogsled
+	_, visualViewport, _, _, cssVisualViewport, _, err := cdppage.GetLayoutMetrics().Do(cdp.WithExecutor(ctx, sess))
+	if err != nil {
+		return 0, 0, 0, fmt.Errorf("getting layout metrics for screenshot: %w", err)
+	}
+
+	// we had a null pointer panic cases, when visualViewport is nil
+	// instead of the erroring out, we fallback to defaults and still try to do a screenshot
+	switch {
+	case cssVisualViewport != nil:
+		visualViewportScale = cssVisualViewport.Scale
+		visualViewportPageX = cssVisualViewport.PageX
+		visualViewportPageY = cssVisualViewport.PageY
+	case visualViewport != nil:
+		visualViewportScale = visualViewport.Scale
+		visualViewportPageX = visualViewport.PageX
+		visualViewportPageY = visualViewport.PageY
+	default:
+		logger.Warnf(
+			"Screenshotter::screenshot",
+			"chrome browser returned nil on page.getLayoutMetrics, falling back to defaults for visualViewport "+
+				"(scale: %v, pageX: %v, pageY: %v)."+
+				"This is non-standard behavior, if possible please report this issue (with reproducible script) "+
+				"to the https://github.com/grafana/xk6-browser/issues/1502.",
+			visualViewportScale, visualViewportPageX, visualViewportPageY,
+		)
+	}
+
+	return visualViewportScale, visualViewportPageX, visualViewportPageY, nil
 }
 
 //nolint:funlen

--- a/common/screenshotter.go
+++ b/common/screenshotter.go
@@ -174,19 +174,30 @@ func (s *screenshotter) screenshot(
 
 	// Add clip region
 	//nolint:dogsled
-	_, visualViewport, _, _, _, _, err := cdppage.GetLayoutMetrics().Do(cdp.WithExecutor(s.ctx, sess))
+	_, _, _, _, visualViewport, _, err := cdppage.GetLayoutMetrics().Do(cdp.WithExecutor(s.ctx, sess))
 	if err != nil {
 		return nil, fmt.Errorf("getting layout metrics for screenshot: %w", err)
 	}
 
+	visualViewportScale := 1.0
+	visualViewportPageX, visualViewportPageY := 0.0, 0.0
+	// we had a null pointer panic cases, when visualViewport is nil
+	// instead of the erroring out, we fallback to defaults and still try to do a screenshot
+	// ideally this case also should be logged
+	if visualViewport != nil {
+		visualViewportScale = visualViewport.Scale
+		visualViewportPageX = visualViewport.PageX
+		visualViewportPageY = visualViewport.PageY
+	}
+
 	if doc == nil {
 		s := Size{
-			Width:  viewport.Width / visualViewport.Scale,
-			Height: viewport.Height / visualViewport.Scale,
+			Width:  viewport.Width / visualViewportScale,
+			Height: viewport.Height / visualViewportScale,
 		}.enclosingIntSize()
 		doc = &Rect{
-			X:      visualViewport.PageX + viewport.X,
-			Y:      visualViewport.PageY + viewport.Y,
+			X:      visualViewportPageX + viewport.X,
+			Y:      visualViewportPageY + viewport.Y,
 			Width:  s.Width,
 			Height: s.Height,
 		}
@@ -194,7 +205,7 @@ func (s *screenshotter) screenshot(
 
 	scale := 1.0
 	if viewport != nil {
-		scale = visualViewport.Scale
+		scale = visualViewportScale
 	}
 	clip = &cdppage.Viewport{
 		X:      doc.X,

--- a/common/screenshotter.go
+++ b/common/screenshotter.go
@@ -13,6 +13,7 @@ import (
 	"github.com/chromedp/cdproto/cdp"
 	"github.com/chromedp/cdproto/emulation"
 	cdppage "github.com/chromedp/cdproto/page"
+	"github.com/grafana/xk6-browser/log"
 )
 
 // ScreenshotPersister is the type that all file persisters must implement. It's job is
@@ -67,10 +68,15 @@ func (f *ImageFormat) UnmarshalJSON(b []byte) error {
 type screenshotter struct {
 	ctx       context.Context
 	persister ScreenshotPersister
+	logger    *log.Logger
 }
 
-func newScreenshotter(ctx context.Context, sp ScreenshotPersister) *screenshotter {
-	return &screenshotter{ctx, sp}
+func newScreenshotter(
+	ctx context.Context,
+	sp ScreenshotPersister,
+	logger *log.Logger,
+) *screenshotter {
+	return &screenshotter{ctx, sp, logger}
 }
 
 func (s *screenshotter) fullPageSize(p *Page) (*Size, error) {
@@ -183,11 +189,19 @@ func (s *screenshotter) screenshot(
 	visualViewportPageX, visualViewportPageY := 0.0, 0.0
 	// we had a null pointer panic cases, when visualViewport is nil
 	// instead of the erroring out, we fallback to defaults and still try to do a screenshot
-	// ideally this case also should be logged
 	if visualViewport != nil {
 		visualViewportScale = visualViewport.Scale
 		visualViewportPageX = visualViewport.PageX
 		visualViewportPageY = visualViewport.PageY
+	} else {
+		s.logger.Warnf(
+			"Screenshotter::screenshot",
+			"chrome browser returned nil on page.getLayoutMetrics, falling back to defaults for visualViewport "+
+				"(scale: %v, pageX: %v, pageY: %v)."+
+				"This is non-standard behavior, if possible please report this issue (with reproducible script) "+
+				"to the https://github.com/grafana/xk6-browser/issues/1502.",
+			visualViewportScale, visualViewportPageX, visualViewportPageY,
+		)
 	}
 
 	if doc == nil {

